### DISCUSSION TITLE
How to use SPRAM on NSPanel ?

### DISCUSSION THREAD

**Discussion by yterza** [arendst/Tasmota#22177] [Q&A] [answered]
Hi everyone,

I'm looking to activate SPRAM on my NSpanel.
https://blakadder.com/nspanel-teardown/ :  we learn here that :

> The brain is an ESP32-D0WD-V3 chip complemented with 4MB Winbond flash and a 16MB PSRAM chip

As I understood, for tasmota PSram is supposed to be on GPIO17/18.

https://community.home-assistant.io/t/sonoff-nspanel-by-itead-smart-scene-wall-switch-based-on-esp32-and-custom-nextion-touch-screen-panel-display-non-pro-variant/332962/1381

> Some guys at [NSPanel HA Blueprint 17](https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/969ff5839fc5f9617dc243631f74f884dbf8460d/components/nspanel_ha_blueprint/__init__.py) figured it out, works with any esphome config :slight_smile:

On the NSpanel, PSram looks to be on GPIO 5/18

How can we set that on tasmota. I tried to configure this GPIO for SPI, but it does nothing.

Thanks for you help

**Comment by sfromis** [comment]
Are you using the specific precompiled NSPanel build, `tasmota32-nspanel`? That build includes a setup meant to handle the pin difference for PSRAM on the device.

Standard Tasmota ESP32 setup for PSRAM is gpio 16,17, but the above is supposed to change this. And don't have any Tasmota config for the relevant pins.

**Reply by yterza** [reply]
I tried with self compiled binary or with tasmota32-nspanel precompiled.

It should appear there ?

![screenshot](https://github.com/user-attachments/assets/d9863723-dc38-4100-9308-f2bf432a4d33)

**Reply by sfromis** [reply]
Indeed, PSRAM info should (if detected) appear right after the "Free Memory" line. And I've seen an old thread where flashing `tasmota32-nspanel` led to Tasmota detecting PSRAM at boot. Maybe you have a message there about detection not working, might be the case for older ESP32 chips.

**Reply by yterza** [reply]
Be there, you mean in the console, at boot time ?
I see no mention to PSRAM

![scrrenshot](https://github.com/user-attachments/assets/3b092dd1-60bc-4cec-80e8-bfc598234d95)

**Reply by sfromis** [reply]
Right, no PSRAM-related messages there. Those 3 `{"Command":"Unknown"}` should probably indicate something else bad or missing, but not about PSRAM. You could enable `Weblog 3` before booting again, to see what's failing.

When it comes to PSRAM not detected, maybe @Jason2866 has some input to whether it is supposed to work with the `nspanel` binary, and why not.

**Reply by yterza** [reply]
Increasing verbosity didn't help. There is no mention to PSRAM during boot sequence. I would expect an error message but I got nothing, which make me thinks that PSRAM is not activated/compiled at all.

**Comment by Jason2866** [comment]
Maybe the changed the board design? Needs investigation which GPIOS are used for PSRAM with your board.

**Reply by yterza** [reply]
I checked, and PSRAM is actually connected to gpio 5/18 as expected.
I tried with ESPHome and PSRAM is available with this framework (but I want to stay with tasmota)

**Reply by Jason2866** [reply]
@yterza Think I found something in Arduino Lib Builder. The sdkconfig setting has changed. Will come up with a test version tomorrow

**Reply by Jason2866** [reply]
@yterza Try this build
[firmware_nspanel.zip](https://github.com/user-attachments/files/17147659/firmware_nspanel.zip)

**Reply by yterza** [reply]
Working like a charm, well done !
How can I compile my own build like this ?

![Capture d’écran du 2024-09-26 20-44-59](https://github.com/user-attachments/assets/89e87034-3c84-4ef7-b069-d0f6dc86af6f)

**Reply by Jason2866** [reply, marked as answer]
Use latest development branch. The fixed framework is already merged. Use the env `tasmota32-nspanel` That's it.

**Reply by yterza** [reply]
That is working as expected, thank you so much !
